cfme/fixtures/virtual_machine.py

```diff
@@ -1,19 +1,18 @@
 """ Fixtures ensuring that a VM/instance is in the specified state for the test
 """
 
-
 import pytest
-from cfme.common.vm import VM
+from cfme.cloud.provider import CloudProvider
 
 
 def _get_vm_obj_if_exists_on_provider(provider, vm_name):
-    vm = VM.factory(vm_name, provider)
-    if not vm.does_vm_exist_on_provider():
-        raise ValueError(
-            "Unable to ensure VM state: "
-            "VM '{}' does not exist on provider '{}'".format(vm_name, provider.key)
-        )
-    return vm
+    if not provider.mgmt.does_vm_exist(vm_name):
+        raise ValueError("Unable to ensure VM state: VM '{}' does not exist on provider '{}'"
+                         .format(vm_name, provider.key))
+    collection = getattr(provider.appliance.collections,
```

Comment:
Yep, but since an appliance doesn't have any collection of infrastructure and cloud vms/instances, I've purposely avoided such a collection definition. This is one of the unfortunate side effects on tests that potentially collect against multiple provider types.

There are probably a few places where the test only collects against InfraProvider or CloudProvider types, and not a combination of the two. If the test wasn't in the `cfme/tests/infrastructure` or `cfme/tests/cloud` folder, I went the cautious route and dynamically picked the collection type, so that if test collection changes we don't get AttributeErrors.

Reply:
OK, so I see this used a lot and I'm still confused on why this couldn't be a bit easier to do repetitively.. 

Since we do this exact step so much shouldn't this live in a helper method somewhere? Something like: `def vm_or_instance_collection_for(provider)` (maybe there's a better name) and everywhere you have these 3 lines you'd now just do `collection = vm_or_instance_collection_for(provider)`. I just don't know where it's best for that helper to live... I guess it would be nice if it could live in something that's already imported in all these fixture/test files ;)

Along similar lines maybe we also have a `instantiate_vm_or_instance(provider, name, template_name)` instead of the `provider.appliance.collections.[infra_vms|cloud_instances].instantiate()` type calls everywhere, and we could call that 1 method in many places?

Reply:
I'll look at putting something in VMCollection, this is a bad pattern to repeat this so often.